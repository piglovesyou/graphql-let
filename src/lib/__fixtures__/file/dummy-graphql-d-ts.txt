/* a7870383395320c75dc806f677ffa7caad7522bc
 * This file is automatically generated by graphql-let.
 */
