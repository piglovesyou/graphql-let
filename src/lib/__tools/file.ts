import copyfiles from 'copyfiles';
import { promises } from 'fs';
import { join as pathJoin, sep } from 'path';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

const crlf = /\r\n/g;
export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}

export const { readFile: _readFile, writeFile, rename } = promises;

export function readFile(file: string) {
  return _readFile(file, 'utf-8').then(normalizeNewLine);
}

export function cleanup(cwd: string, relPaths: string[]) {
  return Promise.all(relPaths.map((rel) => rimraf(pathJoin(cwd, rel))));
}

export function copyDir(
  baseFullDir: string,
  srcRelDir: string,
  destRelDir: string,
): Promise<void> {
  // baseFullDir: /Users/a/b/c
  // srcRelDir: d/e
  // destRelDir: .d/e
  // It generates /Users/a/b/c/.d/e
  if (srcRelDir === destRelDir) throw new Error('Kidding me?');
  const up = pathJoin(baseFullDir).split(sep).length + 1;
  const [relRoot] = destRelDir.split('/');
  return new Promise<void>((resolve, rejects) => {
    copyfiles(
      [pathJoin(baseFullDir, srcRelDir, '**'), pathJoin(baseFullDir, relRoot)],
      { error: true, up, all: true },
      (err) => {
        if (err) return rejects(err);
        resolve();
      },
    );
  });
}

export type AbsFn = (rel: string) => string;

export async function prepareFixtures(
  baseFullDir: string,
  fixtureSrcRelDir: string,
): Promise<[cwd: string, abs: AbsFn]> {
  if (fixtureSrcRelDir.startsWith('..'))
    throw new Error(
      `It doesn't support backward relative paths like ${fixtureSrcRelDir}`,
    );
  const fixtureDestRelDir = '.' + fixtureSrcRelDir;
  const cwd = pathJoin(baseFullDir, fixtureDestRelDir);
  const abs = (relPath: string) => pathJoin(cwd, relPath);
  await copyDir(baseFullDir, fixtureSrcRelDir, fixtureDestRelDir);
  return [cwd, abs];
}
