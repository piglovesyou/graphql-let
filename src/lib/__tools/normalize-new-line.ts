const crlf = /\r\n/g;

export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}
