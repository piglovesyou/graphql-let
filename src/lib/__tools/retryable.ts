function timeout(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export default async function retryable(
  fn: () => Promise<void>,
  interval: number,
  maxTimeout: number,
) {
  if (maxTimeout < 0) throw new Error('never');
  const end = Date.now() + maxTimeout;
  let lastError: Error = new Error('never');
  let n = 0;
  while (Date.now() < end) {
    try {
      await fn();
      console.info(`Succeeded on the ${++n} time.`);
      return;
    } catch (e) {
      console.info(`${++n} times tried. Retrying...`);
      lastError = e;
      await timeout(interval);
    }
  }
  throw lastError;
}
