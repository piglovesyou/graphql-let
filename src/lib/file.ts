import { promises as fsPromises } from 'fs';
import { createReadStream, existsSync } from 'fs';
import glob from 'globby';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

export const { readFile, writeFile } = fsPromises;
export { readFileSync, statSync } from 'fs';

// Erasing old cache in __generated__ on HMR.
// Otherwise the multiple `declare module "*/x.graphql"` are exposed.
export async function removeByPatterns(cwd: string, ...patterns: string[]) {
  const oldFiles = await glob(patterns, {
    cwd,
    absolute: true,
  });
  await Promise.all(oldFiles.map((f) => rimraf(f)));
}

const leadingStringOfGeneratedContent = '/* ';
const hexHashLength = 40;
export function withHash(sourceHash: string, content: string): string {
  return `${leadingStringOfGeneratedContent}${sourceHash}
 * This file is automatically generated by graphql-let. */

${content}`;
}

export function readHash(filePath: string): Promise<string | null> {
  if (!existsSync(filePath)) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const stream = createReadStream(filePath, {
      encoding: 'utf-8',
      highWaterMark: leadingStringOfGeneratedContent.length + hexHashLength,
    });
    stream.on('error', (error) => reject(error));
    stream.on('data', (chunk: string) => {
      const hash = chunk.slice(leadingStringOfGeneratedContent.length);
      if (hash.length !== hexHashLength) return resolve(null);

      resolve(hash);
    });
    stream.read();
  });
}
