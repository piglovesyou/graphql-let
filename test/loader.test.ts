/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { deepStrictEqual, ok, strictEqual } from 'assert';
import glob from 'globby';
import compiler from './__tools/compile';
import { join as pathJoin } from 'path';
import { rimraf } from './__tools/file';

const cwd = pathJoin(__dirname, '__fixtures/loader');

describe('graphql-let/loader', () => {
  beforeEach(async () => {
    await rimraf(pathJoin(cwd, '__generated__'));
  });

  test(
    'generates .tsx and .d.ts',
    async () => {
      const fixture = 'pages/viewer.graphql';
      const stats = await compiler(cwd, fixture, 'node');
      const { 0: actual, length } = stats
        .toJson()
        .modules!.map((m) => m.source)
        .filter(Boolean);

      deepStrictEqual(length, 1);
      expect(actual).toMatchSnapshot();
    },
    60 * 1000,
  );

  test(
    'runs well for simultaneous execution, assuming SSR',
    async () => {
      const expectedTargets: [string, 'node' | 'web'][] = [
        ['pages/viewer.graphql', 'node'],
        ['pages/viewer.graphql', 'web'],
        ['pages/viewer2.graphql', 'node'],
        ['pages/viewer2.graphql', 'web'],
      ];
      const results = await Promise.all(
        expectedTargets.map(([file, target]) => compiler(cwd, file, target)),
      );
      for (const [i, stats] of results.entries()) {
        const [file] = expectedTargets[i];
        const { 0: actual, length } = stats
          .toJson()
          .modules!.map((m) => m.source)
          .filter(Boolean);

        deepStrictEqual(length, 1);
        switch (file) {
          case 'pages/viewer.graphql':
            ok(actual!.includes('export function useViewerQuery('));
            break;
          case 'pages/viewer2.graphql':
            ok(actual!.includes('export function useViewer2Query('));
            break;
        }
      }
      const globResults = await glob('**/*.graphql.d.ts', { cwd });
      strictEqual(globResults.length, 2);
    },
    60 * 1000,
  );
});
