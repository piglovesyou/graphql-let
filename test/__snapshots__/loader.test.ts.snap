// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-let/loader generates .tsx and .d.ts 1`] = `
"function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import gql from 'graphql-tag';
import * as React from 'react';
import * as ApolloReactComponents from '@apollo/react-components';
import * as ApolloReactHooks from '@apollo/react-hooks';
export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    id
    name
    status
  }
}
    \`;
export const ViewerComponent = props => /*#__PURE__*/React.createElement(ApolloReactComponents.Query, _extends({
  query: ViewerDocument
}, props));
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export function useViewerQuery(baseOptions) {
  return ApolloReactHooks.useQuery(ViewerDocument, baseOptions);
}
export function useViewerLazyQuery(baseOptions) {
  return ApolloReactHooks.useLazyQuery(ViewerDocument, baseOptions);
}"
`;
